### ISSUE TITLE
Covariance-related GAT lifetime mismatch

### ISSUE THREAD

**Issue by QuineDot** [rust-lang/rust#89352]
[I tried this code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c44f2a8b6b9fe671e97c3956e24f8d02):

```rust
#![feature(generic_associated_types)]

use core::marker::PhantomData;

pub trait GenAssoc<T> {
    type Iter<'a>;
    fn iter(&self) -> Self::Iter<'_>;
    fn reborrow<'long: 'short, 'short>(iter: Self::Iter<'long>) -> Self::Iter<'short>;
}

pub struct Wrapper<'a, T: 'a, A: GenAssoc<T>> {
    a: A::Iter<'a>,
    _p: PhantomData<T>,
}

impl<'a, T: 'a, A: GenAssoc<T>> GenAssoc<T> for Wrapper<'a, T, A>
where
    A::Iter<'a>: Clone,
{
    type Iter<'b> = ();
    fn iter<'s>(&'s self) -> Self::Iter<'s> {
        let a = A::reborrow::<'a, 's>(self.a.clone());
    }

    fn reborrow<'long: 'short, 'short>(iter: Self::Iter<'long>) -> Self::Iter<'short> {
        ()
    }
}
```

I expected to see this happen: Successful compilation.

Instead, this happened: Errors as follows.
```text

error[E0308]: mismatched types
  --> src/lib.rs:23:13
   |
23 |         let a = A::reborrow::<'a, 's>(self.a.clone());
   |             ^ lifetime mismatch
   |
   = note: expected type `Sized`
              found type `Sized`
note: the lifetime `'s` as defined on the method body at 22:13...
  --> src/lib.rs:22:13
   |
22 |     fn iter<'s>(&'s self) -> Self::Iter<'s> {
   |             ^^
note: ...does not necessarily outlive the lifetime `'a` as defined on the impl at 16:6
  --> src/lib.rs:16:6
   |
16 | impl<'a, T: 'a, A: GenAssoc<T>> GenAssoc<T> for Wrapper<'a, T, A>
   |      ^^

For more information about this error, try `rustc --explain E0308`.
```

Note that it expected and found "type `Sized`", which doesn't make sense.  Additionally, I don't see what is restricting `'s` to outlive `'a`; and indeed, [the error can be worked around](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b8cbf041d30c1868ba21336c2a8a1e71).

### Meta
Playground:
```text
Nightly channel
Build using the Nightly version: 1.57.0-nightly
(2021-09-28 8f8092cc32ec171becef)
```

The [original use-case](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=400026cd1b780a9f85cf36ae74de0e90) tied the lifetime parameter of the GAT to the type in typical fashion.  Discovered in [this URLO thread](https://users.rust-lang.org/t/unable-to-make-code-compile-with-generic-associated-lifetimes/65281/).

@rustbot label  +F-generic_associated_types +A-lifetimes +requires-nightly +A-diagnostics

**Comment by jackh726** [comment]
So, surprisingly, changing `A::Iter<'a>: Clone` to `for<'b> A::Iter<'b>: Clone` makes everything compile file.

<details>
<summary>
Details
</summary>

Somehow, a `<<A as GenAssoc<T>>::Iter<'a> as std::marker::Sized>` is ending up in the `param_env`. Later, we try to prove `<<A as GenAssoc<T>>::Iter<'b> as std::marker::Sized>`, use that predicate, and get that `'a == 'b`.

</details>

**Comment by jackh726** [comment]
Okay, so I've dug into this a little bit. And I thought I would share where I'm at.

First, let's look at the root cause of this problem: the `A::Iter<'a>: Clone` bound. This leads to a `A::Iter<'a>: Sized` bound in the environment. There is also an implicit `for<'x, Y> <X as GenAssoc<Y>>::Iter<'x>: Sized` bound. At some point, we need to prove that `Self::Iter<'s>: Sized`. Because of the [current](https://github.com/rust-lang/rust/blob/e100ec5bc7cd768ec17d75448b29c9ab4a39272b/compiler/rustc_trait_selection/src/traits/select/mod.rs#L1619) rules for what candidates take preference during trait candidate selection. we pick the predicate in the param_env (which, importantly has the specific lifetime `'a`). Note, the implicit bound is more *generally* applicable, but the way the trait selection code works right now means that we select *one* candidate to confirm. So, we eventually unify `'a` with `'s`, and that's a problem.

One potentially solution I had was to just always prefer the projection candidate. Unfortunately, this ran into a problem with the following case:
```rust
fn discr<T, U>(v: T, value: U)
where
    <T as DiscriminantKind>::Discriminant: PartialEq<U>,
{
    assert!(discriminant_value(&v) == value);
}
```

Basically, the param_env gets a `<T as DiscriminantKind>::Discriminant: PartialEq<U>` predicate and the projection candidate is `<T as DiscriminantKind>::Discriminant: PartialEq<<T as DiscriminantKind>::Discriminant>`. We need the former for type checking to succeed. Again though, both candidates are "compatible", and it would just mean that we end up with the constraint that `<T as DiscriminantKind>::Discriminant == U`.

One thing I'm exploring right now is to make candidate preference depend on if it's a `?X: Sized` predicate or not; if so, then we should use the projection candidate (since this will *always* have the same or more generality and doesn't have any parameters) - otherwise, use the param_env candidate. This is a total hack, but might be okay.

**Comment by BoxyUwU** [comment]
I think this isnt solely a gats issue you can reproduce it on stable [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e7028428396d59b5471e3604519f571e) This technically makes reverting the fix a breaking change since the fix hit stable, if you look at the playground you'll see that it compiles on stable but not on latest nightly

**Comment by jackh726** [comment]
uh oh

**Comment by jackh726** [comment]
@craterbot run name=92191_revert mode=check-only start=master#4c60a0ea5b2385d7400df9db1ad04e96f2a4c154 end=master#bdcb6a99e853732f8ec050ae4986aa3af51d44c5

**Comment by danielhenrymantilla** [comment]
> It did break https://github.com/rust-lang/rust/issues/93262 though (uses GATs, but nobody to my knowledge has tried to make a non-GAT repro)

@jackh726 FWIW, neither of the two canonical stable-GAT-workarounds for that snippet regresses: 
  - https://rust.godbolt.org/z/1Y7denM7Y
  - https://rust.godbolt.org/z/EjxPjcM3K

so I strongly suspect this is a GAT-specific regression

**Comment by jackh726** [comment]
Yes, that makes sense: the issue there really comes down to the where clause.

Crater results look clean though. As discussed on Zulip (https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Covariance-related.20GAT.20lifetime.20mismatch.20.2389352/near/281881315), accepting that for beta backport.

**Comment by jackh726** [comment]
This now compiles, since NLL stabilization.

**Comment by jackh726** [comment]
Actually, this was a bug test that got properly moved during stabilization. Closing.
